### PR TITLE
Ruby: add Parameter.getLogicalPosition()

### PR DESCRIPTION
**Pull request by asgerf**
Adds `Parameter.getLogicalPosition()` as a way to get the index of a positional argument, while taking into account that it may be declared after a splat parameter (and therefore have a non-constant logical index).

Commit-by-commit review recommended.

### CODE REVIEW

**Comment by asgerf** [comment]
Not sure if the AST node is the correct place for this, please advice.

Exposing the `ArgumentPosition` and `ParameterPosition` classes could also make for a more general interface. However, I believe the intention is to add synthetic argument/parameter positions in the future, which we'd probably want to keep internal.

File: ruby/ql/lib/codeql/ruby/ast/Parameter.qll [hunk 1]
```diff
@@ -19,6 +19,27 @@ class Parameter extends AstNode, TParameter {
...
+  }
+
+  /**
+   * Gets the zero-based position of this parameter, if it is a plain positional parameter.
+   *
+   * This corresponds to the index of a positional argument that can flow into this parameter.
+   *
+   * Has no result for non-positional parameters, splat parameters, or parameters declared after a splat parameter.
+   */
+  final int getLogicalPosition() {
```
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/ast/Parameter.qll:37]: I think this predicate should not be in the public API .

File: ruby/ql/lib/codeql/ruby/ast/Parameter.qll [hunk 2]
```diff
@@ -19,6 +19,27 @@ class Parameter extends AstNode, TParameter {
...
+      this = c.getParameter([n .. c.getNumberOfParameters() - 1])
+    )
+  }
+
+  /**
+   * Gets the zero-based position of this parameter, if it is a plain positional parameter.
+   *
+   * This corresponds to the index of a positional argument that can flow into this parameter.
+   *
+   * Has no result for non-positional parameters, splat parameters, or parameters declared after a splat parameter.
```
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/ast/Parameter.qll:35]: What about 
```
def foo (a,b=1,c)
end
```
I think in this case `c` does not have a "logical" position either.

File: ruby/ql/lib/codeql/ruby/ast/Parameter.qll (hunk 1 above)
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/ast/Parameter.qll:37]: I also think this should just be a private predicate in `DataFlowPrivate`.

File: ruby/ql/lib/codeql/ruby/ast/Parameter.qll (hunk 2 above)
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ast/Parameter.qll:35]: I see. I was under the impression that the splat parameter actually was the only thing that could offset the index of other positional parameters by a variable amount.

**Comment by asgerf** [comment]
As discussed, I'll close this PR and use `ParameterPosition` directly in the API graph PR instead, so we don't have to commit to anything in the public API.

**Pull request closed without merging**